### PR TITLE
Strict promises

### PR DESCRIPTION
**Pull request by ian-r-rose**
This is an attempt to cut down on async errors in our test suite. It enables two new rules in our `tslint` settings.
1. `'no-floating-promises'`: All promises and async functions *must* be handled. This catches bugs where promises were intended to be chained but were not, or things were intended to be awaited but were not. It *does* force some extra care around "fire-and-forget" promises. In that case, we must consume the result using the `void` operator.
Probably two-thirds of this PR is annontating fire-and-forget promises with `void` to indicate that. It adds some visual noise, and takes some getting used-to, but I think it is overall a good thing. We *should* take extra note of those cases, and sometimes they are not intentional, or are indicative of API awkwardness.
2. `'await-promise'`: All things which are awaited *must* be promises. This should be uncontroversial, but there is one place in the API where we are intentionally awaiting things that may not be promises: `Comms`. In that case, we are awaiting `void | PromiseLike<void>`. @jasongrout has indicated that he may want to revisit that later, but for now I have disabled the linter in those places.

Most of this PR is pretty mechanical. The majority of it is annotating intentional fire-and-forget promises with `void`. A solid chunk of it is changing some functions to return promises instead of `void` when that is more appropriate. There is a handful of places where there were real bugs. I will try to flag those, as well as API changes in the comments.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -50,7 +50,7 @@
     "test:chrome-headless": "lerna run test:chrome-headless --scope \"@jupyterlab/test-*\" --concurrency 1 --stream",
     "test:firefox": "lerna run test:firefox --scope \"@jupyterlab/test-*\" --concurrency 1 --stream",
     "test:ie": "lerna run test:ie --scope \"@jupyterlab/test-*\" --concurrency 1 --stream",
-    "tslint": "tslint --fix -c tslint.json '**/*{.ts,.tsx}'",
+    "tslint": "tslint --fix -c tslint.json --project tsconfigbase.json '**/*{.ts,.tsx}'",
```
Comment (ian-r-rose) [inline package.json]: In order for tslint to reason about types, we need to include the `tsconfig` file.

File: packages/apputils/src/instancetracker.ts [hunk 2]
```diff
@@ -380,7 +386,7 @@ export class InstanceTracker<T extends Widget>
    *
    * @param widget - The widget being saved.
    */
-  save(widget: T): void {
+  async save(widget: T): Promise<void> {
```
Comment (ian-r-rose) [inline packages/apputils/src/instancetracker.ts]: This is an API change. I thought it was appropriate to allow users of `InstanceTracker.save` to know when the saving has completed.

File: packages/apputils/src/instancetracker.ts [hunk 3]
```diff
@@ -320,7 +326,7 @@ export class InstanceTracker<T extends Widget>
    */
   inject(widget: T): void {
     Private.injectedProperty.set(widget, true);
-    this.add(widget);
+    void this.add(widget);
```
Comment (ian-r-rose) [inline packages/apputils/src/instancetracker.ts]: I initially made this async, but the note indicates that injected widgets are not saved, so we don't need to wait on that.

File: packages/docmanager/src/widgetmanager.ts [hunk 4]
```diff
@@ -257,7 +257,7 @@ export class DocumentWidgetManager implements IDisposable {
    *
    * @param widget - The target widget.
    */
-  protected setCaption(widget: Widget): void {
+  protected setCaption(widget: Widget): Promise<void> {
```
Comment (ian-r-rose) [inline packages/docmanager/src/widgetmanager.ts]: Another API change.

File: packages/filebrowser/src/browser.ts [hunk 5]
```diff
@@ -200,8 +205,8 @@ export class FileBrowser extends Widget {
   /**
    * Download the currently selected item(s).
    */
-  download(): void {
-    this._listing.download();
+  download(): Promise<void> {
```
Comment (ian-r-rose) [inline packages/filebrowser/src/browser.ts]: Another API change: allow users to hook into when the download promise is resolved.

**Comment by blink1073** [comment]
This is amazing, @ian-r-rose l, nice work!

**Comment by vidartf** [comment]
As we've had hard to track bugs/problems with this, I say the benefits outweigh the "noise". And I would disagree to calling it noise, as it has *actual meaning*.

> *Explicit is better than implicit*
>  -- The Zen of Python

**Pull request merged**